src/SharpLearning.Optimization/BayesianOptimizer.cs

```diff
@@ -234,68 +262,69 @@ public OptimizerResult[] Optimize(Func<double[], OptimizerResult> functionToMini
                             bestParameterSet = set;
                         }
                     }
-                }
+                });
             }
 
-            var lastSet = new double[m_parameters.Length];
+            var previousSets = new ConcurrentBag<double[]>();
             for (int iteration = 0; iteration < m_iterations; iteration++)
             {
                 // fit model
-                var observations = parameterSets.ToF64Matrix();
+                var observations = parameterSets.ToList().ToF64Matrix();
                 var targets = parameterSetScores.ToArray();
                 var model = m_learner.Learn(observations, targets);
 
                 var bestScore = parameterSetScores.Min();
                 var candidates = FindNextCandidates(model, bestScore);
-                
+
                 var first = true;
 
-                foreach (var candidate in candidates)
+                Parallel.ForEach(candidates, m_parallelOptions, candidate =>
                 {
                     var parameterSet = candidate.ParameterSet;
 
-                    if (Equals(lastSet, parameterSet) && !first)
+                    // skip evaluation if parameters have not changed unless explicitly allowed
+                    if (m_allowMultipleEvaluations || first || !Contains(previousSets, parameterSet))
```

Comment:
@jameschch I am in doubt if it provides value to allow multiple evaluations, to support random behavior in the `function  ToMiminize`. Can you elaborate a bit on when this is useful? 

My take on this would be if the function being minimized is random, then it does not make much sense to try an optimize parameters for it, since there is no guarantees that the parameters will yield the same results when used again.

Reply:
The specific use I have for this is for parameters on QLearning models. Depending on the exploration rate, a certain proportion of actions will be determined randomly. This can result in significant swings in the result until the action value function converges.

Reply:
@jameschch Ok, that might make sense then. I am no expert in QLearning, so please bear with me here :-). The certain proportion of actions that will be determined randomly, is it not possible to seed these, to make them "deterministic" random? Meaning that you will get the same "random" behavior given the same seed?

Reply:
I am also not an expert. Regardless, there may be different rewards associated with taking different actions given identical states. If you always choose the same action, you never encounter the entire range of possible rewards and therefore cannot learn the optimal strategy.

Reply:
@jameschch Since it can be switched on and off let's leave it in for now. I will complete the pull request, but delay the release a bit since there are some things I want to test locally first.